service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/ClonePolicyAttributesStep.java

```diff
@@ -0,0 +1,105 @@
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.StepStatus;
+import bio.terra.stairway.exception.RetryException;
+import bio.terra.workspace.amalgam.tps.TpsApiDispatch;
+import bio.terra.workspace.generated.model.ApiTpsComponent;
+import bio.terra.workspace.generated.model.ApiTpsObjectType;
+import bio.terra.workspace.generated.model.ApiTpsPaoCreateRequest;
+import bio.terra.workspace.generated.model.ApiTpsPaoGetResult;
+import bio.terra.workspace.generated.model.ApiTpsPaoSourceRequest;
+import bio.terra.workspace.generated.model.ApiTpsPaoUpdateResult;
+import bio.terra.workspace.generated.model.ApiTpsPolicyInputs;
+import bio.terra.workspace.generated.model.ApiTpsUpdateMode;
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import bio.terra.workspace.service.resource.exception.PolicyConflictException;
+import bio.terra.workspace.service.workspace.flight.WorkspaceFlightMapKeys;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ClonePolicyAttributesStep implements Step {
+  private static final Logger logger = LoggerFactory.getLogger(ClonePolicyAttributesStep.class);
+
+  private final UUID sourceWorkspaceId;
+  private final UUID destinationWorkspaceId;
+  private final AuthenticatedUserRequest userRequest;
+  private final TpsApiDispatch tpsApiDispatch;
+
+  public ClonePolicyAttributesStep(
+      UUID sourceWorkspaceId,
+      UUID destinationWorkspaceId,
+      AuthenticatedUserRequest userRequest,
+      TpsApiDispatch tpsApiDispatch) {
+    this.sourceWorkspaceId = sourceWorkspaceId;
+    this.destinationWorkspaceId = destinationWorkspaceId;
+    this.userRequest = userRequest;
+    this.tpsApiDispatch = tpsApiDispatch;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    BearerToken token = new BearerToken(userRequest.getRequiredToken());
+
+    // Create PAOs if they don't exist; catch TPS exceptions and retry
+    try {
+      createPaoIfNotExist(token, sourceWorkspaceId);
+      createPaoIfNotExist(token, destinationWorkspaceId);
+    } catch (Exception ex) {
+      logger.info("Attempt to create a PAO for workspace failed", ex);
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, ex);
+    }
+
+    // Save the destination attributes so we can restore them if the flight fails
+    ApiTpsPaoGetResult destinationPao = tpsApiDispatch.getPao(token, destinationWorkspaceId);
+    ApiTpsPolicyInputs destinationAttributes = destinationPao.getAttributes();
+    context.getWorkingMap().put(WorkspaceFlightMapKeys.POLICIES, destinationAttributes);
+
+    ApiTpsPaoSourceRequest request =
+        new ApiTpsPaoSourceRequest()
+            .sourceObjectId(sourceWorkspaceId)
+            .updateMode(ApiTpsUpdateMode.FAIL_ON_CONFLICT);
+
+    ApiTpsPaoUpdateResult result = tpsApiDispatch.mergePao(token, destinationWorkspaceId, request);
```

Comment:
I haven't looked at the TPS implmenetation but just to confirm: is it always a no-op to merge a PAO from a workspace into itself? For resource-level clone operations, I think it's possible to clone within a workspace

Reply:
Yes. There wasn't originally, but when I implemented the per-resource clone and looked at all of the tests, I realized we did it all of the time! So now it is a no-op to merge a PAO to itself.